src/types.ts

```diff
@@ -50,25 +60,42 @@ export type Flow = {
 
 export interface IndexedNode extends Node {
   id: string;
+  parentId: string | null; // null if it is the first node
```

Comment:
Great point, this could be represented more precisely in types. I do think this is nice but might be more overhead than is helpful at this point, like you mentioned. Once we've got a few use cases in, it'd be worth trying this out to see if it simplifies or complicates things.